R/txrange.R

```diff
@@ -0,0 +1,265 @@
+#' Assigns each read class geneScore and txScore
+#' @param se summerized experiment object with read classes/ranges
+#' @param genomeSequence genomeSequence
+#' @param annotations GRangesList of annotations
+txrange.scoreReadClasses = function(se, genomeSequence, annotations, 
+                                    min.readCount = 2){
+    options(scipen = 999)
+    se = addRowData(se, genomeSequence, annotations)
+    thresholdIndex = which(rowData(se)$readCount
+        >=min.readCount)
+    rowData(se)$geneScore = getGeneScore(se, thresholdIndex)
+    rowData(se)$txScore = getTranscriptScore(se, thresholdIndex)
+    return(se)
+}
+
+#' calculates labels and features used in model generation
+addRowData = function(se, genomeSequence, annotations){
+  rowData(se)$numExons <- elementNROWS(rowRanges(se))
+  rowData(se)$equal = isReadClassEqual(rowRanges(se), annotations)
+  rowData(se)$compatible = isReadClassCompatible(rowRanges(se), annotations)
+  rowData(se)$GENEID = assignGeneIds(rowRanges(se), annotations)
+  rowData(se)$novel = grepl("gene.", 
+      rowData(se)$GENEID)
+  se = calculateGeneProportion(se)
+  se = countPolyATerminals(se, genomeSequence)
+  return(se)
+}
+
+#' % of a genes read counts assigned to each read class
+calculateGeneProportion = function(resultOutput){
+  countsTBL <- as_tibble(rowData(resultOutput)$readCount) %>%
+    mutate(geneId = rowData(resultOutput)$GENEID) %>%
+    group_by(geneId) %>%
+    mutate_at(vars(-geneId), .funs = sum) %>%
+    ungroup() %>%
+    dplyr::select(-geneId)
+  geneReadProp <- rowData(resultOutput)$readCount / countsTBL
+  rowData(resultOutput)$geneReadProp = unlist(geneReadProp)
+  return(resultOutput)
+}
+
+#' checks to see if a read classes intron junctions fully matches an annotation
+isReadClassEqual = function(query, subject){
+    olapEqual = findOverlaps(cutStartEndFromGrangesList(query),
+      cutStartEndFromGrangesList(subject), ignore.strand = F, type = 'equal',
+      select= 'first')
+    equal = !is.na(olapEqual)
+    return(equal)
+}
+
+#' returns number of ref anno each read class is a subset of
+isReadClassCompatible =  function(query, subject){
+  compatible = rep(0, length(query))
+  
+  olap = findOverlaps(cutStartEndFromGrangesList(query),
+                      cutStartEndFromGrangesList(subject), ignore.strand = F, type = 'within')
+
+  query <- query[queryHits(olap)]
+  subject <- subject[subjectHits(olap)]
+  splice <- myGaps(query)
+
+  qrng <- ranges(query)
+  srng <- ranges(subject)
+  sprng <- ranges(splice)
+  
+  #calculates if query is a subset of subject
+  bnds <- elementNROWS(GenomicRanges::setdiff(qrng, srng)) == 0L
+  splc <- elementNROWS(GenomicRanges::intersect(srng, sprng)) == 0L
+  
+  #count number of compatible matches
+  counts = by(bnds & splc, queryHits(olap), sum)
```

Comment:
change is faster by ~500 micro seconds.

Reply:
Implemented in 8936cbd

Reply:
what about this here? will make it consistent with how equal is calculated?
`outData$compatible <- countQueryHits(olap[comp])`

Reply:
i.e. counts is not neede here I think

Reply:
Sorry I think I am missing some context to help me understand what you mean by those two comments. Are you referencing another commit/comment?

If you mean why isn't $compatible a boolean vector but a numeric vector, its because I need to know the number of compatible read classes to determine if its a subset or not for the gene model

Reply:
no much simpler suggestion just coding style, instead of `counts <-X; outData$compatible <- counts; ` we can do `outData$compatible <- X` will remove some lines

Reply:
ahh I see! yes thats a logical change.  changed in 5e19452